kustomize-functions/validate-asm/image/main.go

```diff
@@ -0,0 +1,347 @@
+// SPDX-License-Identifier: Apache-2.0
+
+// Package main implements an injection function for resource reservations and
+// is run with `kustomize config run -- DIR/`.
+package main
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+
+	multierr "github.com/hashicorp/go-multierror"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const (
+	loggingServiceValue = "logging.googleapis.com/kubernetes"
+	monitoringServiceValue = "monitoring.googleapis.com/kubernetes"
+)
+
+var supportedReleaseChannels = []string{"REGULAR", "RAPID", "STABLE"}
+var nodeVersionRegex = regexp.MustCompile(`(\d+).(\d+).(\d+)-gke.(\d+)`)
+var n1CustomMachineTypeRegex = regexp.MustCompile(`(\w+)-(\d+)-(\d+)`)
+var customMachineTypeRegex = regexp.MustCompile(`(\w+)-(\w+)-(\d+)-(\d+)`)
+var machineTypeRegex = regexp.MustCompile(`(\w+)-(\w+)-(\d+)`)
+
+func main() {
+	rw := &kio.ByteReadWriter{Reader: os.Stdin, Writer: os.Stdout, KeepReaderAnnotations: true}
+	p := kio.Pipeline{
+		Inputs:  []kio.Reader{rw},       // read the inputs into a slice
+		Filters: []kio.Filter{filter{}}, // run the enableASM into the inputs
+		Outputs: []kio.Writer{rw}}       // copy the inputs to the output
+	if err := p.Execute(); err != nil {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		os.Exit(1)
+	}
+}
+
+// filter implements kio.Filter
+type filter struct{}
+
+// Filter injects new filters into container cluster and nodepool.
+func (filter) Filter(in []*yaml.RNode) ([]*yaml.RNode, error) {
+	var errList []error
+
+	for _, r := range in {
+		if errs := validate(r); len(errs) != 0 {
+			errList = append(errList, errs...)
+		}
+	}
+	if errs := multierr.Append(nil, errList...); errs != nil {
+		return nil, errs
+	}
+	return in, nil
+}
+
+// https://cloud.google.com/service-mesh/docs/gke-install-existing-cluster#setting_up_your_project
+func validate(r *yaml.RNode) []error {
+
+	var errList []error
+
+	meta, err := r.GetMeta()
+	if err != nil {
+		errList = append(errList, err)
+		return errList
+	}
+
+	if meta.Kind == "ContainerCluster" {
+
+		// validate if Cloud Monitoring and Logging are enabled
+		if err := validateNodeValue(r, meta, loggingServiceValue, "spec", "loggingService"); err != nil {
+			errList = append(errList, err)
+		}
+		if err := validateNodeValue(r, meta, monitoringServiceValue, "spec", "monitoringService"); err != nil {
+			errList = append(errList, err)
+		}
+
+		// validate if Workload Identity is enabled
+		if _, err := validateNodeExists(r, meta, "spec", "workloadIdentity", "identityNamespace"); err != nil {
+			errList = append(errList, err)
+		}
+
+		// validate if mesh_id lable is set
+		if _, err := validateNodeExists(r, meta, "spec", "labels", "mesh_id"); err != nil {
+			errList = append(errList, err)
+		}
+
+		// validate release channel
+		if err := validateNodeValueIn(r, meta, supportedReleaseChannels,"spec", "releaseChannel", "channel"); err != nil {
```

Comment:
It seems that `spec.releaseChannel` is already a required field.

Reply:
yes, if `spec.releaseChannel` node is missing, it also returns an error with a missing node.